File: frontend/src/services/bookingService.js
import axiosClient from '../axiosConfig/axiosConfig';

const bookingService = {
  // Tạo booking mới
  createBooking: async (bookingData) => {
    try {
      console.log('Creating booking:', bookingData);
      const response = await axiosClient.post('/bookings', bookingData);
      console.log('✅ Booking created:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Error creating booking:');
      console.error('Status:', error.response?.status);
      console.error('Error data:', error.response?.data);
      console.error('Request data:', bookingData);
      throw error;
    }
  },

  // Kiểm tra availability
  checkAvailability: async (courtId, playDate, startTime, endTime) => {
    try {
      const params = {
        courtId: parseInt(courtId),
        playDate,
        startTime,
        endTime
      };
      console.log('🔍 Checking availability:', params);
      console.log('📤 Full URL:', `/bookings/check-availability?courtId=${courtId}&playDate=${playDate}&startTime=${startTime}&endTime=${endTime}`);

      const response = await axiosClient.get('/bookings/check-availability', { params });
      console.log('✅ Availability response:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Error checking availability:');
      console.error('Status:', error.response?.status);
      console.error('Data:', error.response?.data);
      console.error('Headers:', error.response?.headers);
      console.error('Request params:', error.config?.params);
      throw error;
    }
  },

  // Tính giá
  calculatePrice: async (courtId, playDate, startTime, endTime) => {
    try {
      console.log('💰 Calculating price:', { courtId, playDate, startTime, endTime });
      const response = await axiosClient.get('/bookings/calculate-price', {
        params: { courtId, playDate, startTime, endTime }
      });
      console.log('✅ Price response:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Error calculating price:', error.response?.data || error.message);
      throw error;
    }
  },

  // Lấy thông tin booking
  getBooking: async (bookingId) => {
    try {
      const response = await axiosClient.get(`/bookings/${bookingId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching booking:', error);
      throw error;
    }
  },

  // Lấy bookings của user
  getUserBookings: async (userId, fromDate, toDate) => {
    try {
      const response = await axiosClient.get(`/bookings/user/${userId}`, {
        params: { fromDate, toDate }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching user bookings:', error);
      throw error;
    }
  },

  // Hủy booking
  cancelBooking: async (bookingId, reason) => {
    try {
      const response = await axiosClient.post(`/bookings/${bookingId}/cancel`, null, {
        params: { reason }
      });
      return response.data;
    } catch (error) {
      console.error('Error canceling booking:', error);
      throw error;
    }
  },

  // Lấy available slots
  getAvailableSlots: async (courtId, date) => {
    try {
      const response = await axiosClient.get(`/bookings/court/${courtId}/available-slots`, {
        params: { date }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching available slots:', error);
      throw error;
    }
  },

  // Lấy bookings của sân theo ngày
  getCourtBookings: async (courtId, date) => {
    try {
      const response = await axiosClient.get(`/bookings/court/${courtId}`, {
        params: { date }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching court bookings:', error);
      throw error;
    }
  },

  // Cập nhật trạng thái booking
  updateStatus: async (bookingId, status) => {
    try {
      const response = await axiosClient.patch(`/bookings/${bookingId}/status`, null, {
        params: { status }
      });
      return response.data;
    } catch (error) {
      console.error('Error updating booking status:', error);
      throw error;
    }
  },

  // Gia hạn booking
  extendBooking: async (bookingId, extendData) => {
    try {
      const response = await axiosClient.post(`/bookings/${bookingId}/extend`, extendData);
      return response.data;
    } catch (error) {
      console.error('Error extending booking:', error);
      throw error;
    }
  },
  // User tự hủy/yêu cầu hủy booking
  cancelMyBooking: async (bookingId, reason) => {
    try {
      const response = await axiosClient.post(`/bookings/my-bookings/${bookingId}/cancel`, { reason });
      return response.data;
    } catch (error) {
      console.error('Error canceling my booking:', error);
      throw error;
    }
  },

  // Admin duyệt hủy booking
  approveCancellation: async (bookingId) => {
    try {
      const response = await axiosClient.post(`/bookings/${bookingId}/cancellation/approve`);
      return response.data;
    } catch (error) {
      console.error('Error approving cancellation:', error);
      throw error;
    }
  },

  // Admin từ chối hủy booking
  rejectCancellation: async (bookingId) => {
    try {
      const response = await axiosClient.post(`/bookings/${bookingId}/cancellation/reject`);
      return response.data;
    } catch (error) {
      console.error('Error rejecting cancellation:', error);
      throw error;
    }
  }
};

export default bookingService;

File: backend/src/main/java/com/badminton/booking/service/BookingService.java
package com.badminton.booking.service;

import com.badminton.booking.dto.booking.*;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.List;

/**
 * Service xử lý booking sân - KHÔNG dùng slot, chỉ dùng startTime-endTime
 */
public interface BookingService {

    /**
     * Kiểm tra overlap booking
     * Công thức: (start < existingEnd + buffer) && (end > existingStart)
     */
    AvailabilityResponse checkAvailability(Integer courtId, LocalDate playDate,
            LocalTime startTime, LocalTime endTime);

    /**
     * Tính giá booking - tự động tách theo nhiều khung giá
     * Ví dụ: 07:00-10:00 = (07:00-08:00 x 60k) + (08:00-10:00 x 80k) = 220k
     */
    PriceCalculationResponse calculatePrice(Integer courtId, LocalDate playDate,
            LocalTime startTime, LocalTime endTime);

    /**
     * Tạo booking mới
     */
    BookingResponse createBooking(Integer userId, CreateBookingRequest request);

    /**
     * Gia hạn booking - chỉ cho phép nếu giờ tiếp theo còn trống
     */
    BookingResponse extendBooking(Integer bookingId, ExtendBookingRequest request);

    /**
     * Tính phí overtime khi khách về muộn
     */
    OvertimeResponse calculateOvertimeFee(Integer bookingId, LocalTime actualEndTime);

    /**
     * Hoàn thành booking và tính phí overtime nếu có
     */
    BookingResponse completeBooking(Integer bookingId, LocalTime actualEndTime);

    /**
     * Lấy thông tin booking
     */
    BookingResponse getBooking(Integer bookingId);

    /**
     * Cập nhật trạng thái booking
     */
    BookingResponse updateBookingStatus(Integer bookingId, String newStatus);

    /**
     * Hủy booking
     */
    BookingResponse cancelBooking(Integer bookingId, String reason);

    /**
     * User tự hủy booking của mình (chỉ khi đang PENDING)
     */
    BookingResponse cancelMyBooking(Integer userId, Integer bookingId, String reason);

    /**
     * Admin duyệt yêu cầu hủy booking
     */
    BookingResponse approveCancellation(Integer bookingId);

    /**
     * Admin từ chối yêu cầu hủy booking (trả về CONFIRMED)
     */
    BookingResponse rejectCancellation(Integer bookingId);

    /**
     * Lấy danh sách booking của user
     */
    List<BookingResponse> getUserBookings(Integer userId, LocalDate fromDate, LocalDate toDate);

    /**
     * Lấy tất cả booking của user với filter tùy chọn
     */
    List<BookingResponse> getUserBookingsWithFilter(Integer userId, LocalDate fromDate, LocalDate toDate,
            String status, String paymentStatus);

    /**
     * Lấy danh sách booking của sân trong ngày
     */
    List<BookingResponse> getCourtBookings(Integer courtId, LocalDate date);

    /**
     * Check-in booking (chuyển sang PLAYING)
     */
    BookingResponse checkIn(Integer bookingId);

    /**
     * Lấy các khung giờ trống của sân trong ngày
     */
    List<AvailabilityResponse.AvailableSlot> getAvailableSlots(Integer courtId, LocalDate date);

    /**
     * Lấy tất cả booking với filter và phân trang (dành cho admin)
     */
    BookingPageResponse getAllBookingsWithFilter(Integer courtId, String status, String paymentStatus,
            LocalDate fromDate, LocalDate toDate, int page, int size);

    /**
     * Early Release - Mở slot do khách về sớm
     * Biến thời gian còn dư của booking thành slot rảnh có thể bán lại
     *
     * Rules:
     * - Booking phải COMPLETED
     * - actualEndTime < endTime
     * - Thời gian dư >= MIN_BOOKABLE_MINUTES (30 phút)
     * - Slot chưa từng được release trước đó
     */
    EarlyReleaseResponse releaseEarly(Integer bookingId, String releasedBy);

    /**
     * Admin tạo booking cho khách walk-in (khách đến sân trực tiếp, không đặt
     * trước)
     *
     * Đặc điểm:
     * - Cho phép tạo booking từ thời gian quá khứ (trong giới hạn
     * MAX_PAST_BOOKING_DAYS)
     * - Tự động set trạng thái PLAYING và ghi nhận thời gian check-in
     * - Có thể tạo cho guest (không cần user đăng ký)
     *
     * @param request       Thông tin booking walk-in
     * @param adminUsername Username của admin tạo booking
     * @return BookingResponse
     */
    BookingResponse createWalkInBooking(WalkInBookingRequest request, String adminUsername);
}
